# Word Corrector Project with Natural Language Processing (NLP)

The aim of this project is to develop a word corrector using Natural Language Processing (NLP) according to the grammar of the Portuguese language (PT-BR).

# Loading Data

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
with open("/content/drive/MyDrive/Área de Estudos/Alura/Formações/Python para Data Science/Versao_1/python_database/curso7-corretor-nlp/artigos.txt", "r") as f:
    articles = f.read()

print(articles[:500])




imagem 

Temos a seguinte classe que representa um usuário no nosso sistema:

java

Para salvar um novo usuário, várias validações são feitas, como por exemplo: Ver se o nome só contém letras, [**o CPF só números**] e ver se o usuário possui no mínimo 18 anos. Veja o método que faz essa validação:

java 

Suponha agora que eu tenha outra classe, a classe `Produto`, que contém um atributo nome e eu quero fazer a mesma validação que fiz para o nome do usuário: Ver se só contém letras. E aí? Vou


# Tokenization

In [5]:
len(articles)

2605046

In [6]:
len("Olá")

3

In [7]:
example_text = "Hello! How is it going?"
tokens = example_text.split()

In [8]:
print(len(tokens))

5


In [9]:
print(tokens)

['Hello!', 'How', 'is', 'it', 'going?']


In [10]:
import nltk
nltk.download('punkt')

separated_words = nltk.tokenize.word_tokenize(example_text)                     # Separating my tokens to remove special characters and other unnecessary records

print(separated_words)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


['Hello', '!', 'How', 'is', 'it', 'going', '?']


In [11]:
print(separated_words)

['Hello', '!', 'How', 'is', 'it', 'going', '?']


In [12]:
len(separated_words)

7

In [13]:
'./'.isalpha()

False

In [14]:
def separate_words(token_list):                                                 # Function that returns only the words without special characters
    word_list = []
    for token in token_list:
        if token.isalpha():                                                     # isalpha(): checks if all characters in the token are alphabetic, if true, it's stored in the variable
            word_list.append(token)
    return word_list

separate_words(separated_words)

['Hello', 'How', 'is', 'it', 'going']

In [15]:
token_list = nltk.tokenize.word_tokenize(articles)
word_list = separate_words(token_list)

print(f"The number of words is {len(word_list)}")

The number of words is 403104


In [16]:
print(word_list[:10])

['imagem', 'Temos', 'a', 'seguinte', 'classe', 'que', 'representa', 'um', 'usuário', 'no']


In [17]:
def normalization(word_list):
    normalized_list = []
    for word in word_list:
        normalized_list.append(word.lower())                                    # Standardizing words to all be lowercase
    return normalized_list

normalized_list = normalization(word_list)
print(normalized_list[:10])

['imagem', 'temos', 'a', 'seguinte', 'classe', 'que', 'representa', 'um', 'usuário', 'no']


In [18]:
len(set(normalized_list))                                                       # Removing duplicate words

18465

# Word Corrector

### Testing the Corrector

In [19]:
list = "lgica"
(list[:1],list[1:])

('l', 'gica')

In [21]:
example_word = "lgica"

# Inserting letters into our example word
def insert_letters(slices):
    new_words = []
    letters = 'abcdefghijklmnopqrstuvwxyzáâàãéêèẽíîìĩóôõòúûùũç'
    for left, right in slices:                                                  # Left == left // Right == right
        for letter in letters:                                                  # List of words with all possible letters
            new_words.append(left + letter + right)                             # (L + [letter] + gica)
    return new_words

# Generating different word possibilities
def word_generator(word):
    slices = []
    for i in range(len(word)+1):
        slices.append((word[:i], word[i:]))                                     # Inserting slices into the word generator
    generated_words = insert_letters(slices)                                    # Inserting letters into the word generator
    return generated_words

generated_words = word_generator(example_word)
print(generated_words)
print(example_word)

['algica', 'blgica', 'clgica', 'dlgica', 'elgica', 'flgica', 'glgica', 'hlgica', 'ilgica', 'jlgica', 'klgica', 'llgica', 'mlgica', 'nlgica', 'olgica', 'plgica', 'qlgica', 'rlgica', 'slgica', 'tlgica', 'ulgica', 'vlgica', 'wlgica', 'xlgica', 'ylgica', 'zlgica', 'álgica', 'âlgica', 'àlgica', 'ãlgica', 'élgica', 'êlgica', 'èlgica', 'ẽlgica', 'ílgica', 'îlgica', 'ìlgica', 'ĩlgica', 'ólgica', 'ôlgica', 'õlgica', 'òlgica', 'úlgica', 'ûlgica', 'ùlgica', 'ũlgica', 'çlgica', 'lagica', 'lbgica', 'lcgica', 'ldgica', 'legica', 'lfgica', 'lggica', 'lhgica', 'ligica', 'ljgica', 'lkgica', 'llgica', 'lmgica', 'lngica', 'logica', 'lpgica', 'lqgica', 'lrgica', 'lsgica', 'ltgica', 'lugica', 'lvgica', 'lwgica', 'lxgica', 'lygica', 'lzgica', 'lágica', 'lâgica', 'làgica', 'lãgica', 'légica', 'lêgica', 'lègica', 'lẽgica', 'lígica', 'lîgica', 'lìgica', 'lĩgica', 'lógica', 'lôgica', 'lõgica', 'lògica', 'lúgica', 'lûgica', 'lùgica', 'lũgica', 'lçgica', 'lgaica', 'lgbica', 'lgcica', 'lgdica', 'lgeica', 'lgfica',

In [22]:
def corrector(word):                                                            # Function that generates the word corrector
    generated_words = word_generator(word)
    correct_word = max(generated_words, key=probability)                        # The correct word is defined by the word with the highest probability among the generated words
    return correct_word

In [23]:
frequency = nltk.FreqDist(normalized_list)                                      # Calculating the frequency of each word
total_words = len(normalized_list)
frequency.most_common(10)

[('de', 15502),
 ('o', 14056),
 ('que', 12230),
 ('a', 11099),
 ('e', 10501),
 ('para', 7710),
 ('um', 6368),
 ('é', 5899),
 ('uma', 5220),
 ('do', 5124)]

In [24]:
def probability(generated_word):                                                # Function for word probability to analyze their frequency whenever needed
    return frequency[generated_word] / total_words

probability("logic")

0.0

In [25]:
probability("lógica")

0.00023815194093831864

In [26]:
corrector(example_word)

'lógica'

In [28]:
# Importing a new file to perform a word correction test
def create_test_data(file_name):
    test_word_list = []
    f = open(file_name, "r")
    for line in f:                                                              # Creates a list which separates the lines from my database into two fields, one with correctly spelled words and the other not
        correct, wrong = line.split()
        test_word_list.append((correct, wrong))
    f.close()
    return test_word_list

test_list = create_test_data("/content/drive/MyDrive/Área de Estudos/Alura/Formações/Python para Data Science/Versao_1/python_database/curso7-corretor-nlp/palavras.txt")
test_list

[('podemos', 'pyodemos'),
 ('esse', 'esje'),
 ('já', 'jrá'),
 ('nosso', 'nossov'),
 ('são', 'sãêo'),
 ('dos', 'dosa'),
 ('muito', 'muifo'),
 ('imagem', 'iômagem'),
 ('sua', 'ósua'),
 ('também', 'tambéùm'),
 ('ele', 'eme'),
 ('fazer', 'èazer'),
 ('temos', 'temfs'),
 ('essa', 'eàssa'),
 ('quando', 'quaôdo'),
 ('vamos', 'vamvos'),
 ('sobre', 'hsobre'),
 ('java', 'sjava'),
 ('das', 'daõs'),
 ('agora', 'agorah'),
 ('está', 'eòtá'),
 ('cada', 'céda'),
 ('mesmo', 'zmesmo'),
 ('nos', 'noâ'),
 ('forma', 'fobma'),
 ('seja', 'sejéa'),
 ('então', 'enêão'),
 ('criar', 'èriar'),
 ('código', 'cóeigo'),
 ('caso', 'casío'),
 ('exemplo', 'áexemplo'),
 ('tem', 'tĩem'),
 ('usuário', 'usuárôio'),
 ('dados', 'dfados'),
 ('python', 'pgthon'),
 ('nossa', 'nossah'),
 ('além', 'alémè'),
 ('assim', 'asõim'),
 ('ter', 'teb'),
 ('até', 'atĩ'),
 ('bem', 'âem'),
 ('design', 'desigen'),
 ('trabalho', 'trabalàho'),
 ('foi', 'foo'),
 ('apenas', 'apenaũ'),
 ('empresa', 'empresà'),
 ('valor', 'valíor'),
 ('será', 'serr')

In [29]:
def evaluator(tests):
    word_count = len(tests)
    correct_count = 0
    for correct, wrong in tests:                                                # Returns the count of correctly corrected words based on the correct word
        corrected_word = corrector(wrong)
        if corrected_word == correct:
            correct_count += 1
    accuracy_rate = round(correct_count * 100 / word_count, 2)                  # Rounds the evaluator's result to two decimal places
    print(f"{accuracy_rate}% of {word_count} words")

evaluator(test_list)

1.08% of 186 words


In [30]:
def deleting_characters(slices):
    new_words = []
    for left, right in slices:
        new_words.append(left + right[1:])
    return new_words

In [31]:
def word_generator(word):
    slices = []
    for i in range(len(word)+1):
        slices.append((word[:i], word[i:]))
    generated_words = insert_letters(slices)                                    # Inserts possible letters to generate the corrected word
    generated_words += deleting_characters(slices)                              # Returns the generated words with removed characters
    return generated_words
example_word = "lóigica"
generated_words = word_generator(example_word)
print(generated_words)

['alóigica', 'blóigica', 'clóigica', 'dlóigica', 'elóigica', 'flóigica', 'glóigica', 'hlóigica', 'ilóigica', 'jlóigica', 'klóigica', 'llóigica', 'mlóigica', 'nlóigica', 'olóigica', 'plóigica', 'qlóigica', 'rlóigica', 'slóigica', 'tlóigica', 'ulóigica', 'vlóigica', 'wlóigica', 'xlóigica', 'ylóigica', 'zlóigica', 'álóigica', 'âlóigica', 'àlóigica', 'ãlóigica', 'élóigica', 'êlóigica', 'èlóigica', 'ẽlóigica', 'ílóigica', 'îlóigica', 'ìlóigica', 'ĩlóigica', 'ólóigica', 'ôlóigica', 'õlóigica', 'òlóigica', 'úlóigica', 'ûlóigica', 'ùlóigica', 'ũlóigica', 'çlóigica', 'laóigica', 'lbóigica', 'lcóigica', 'ldóigica', 'leóigica', 'lfóigica', 'lgóigica', 'lhóigica', 'lióigica', 'ljóigica', 'lkóigica', 'llóigica', 'lmóigica', 'lnóigica', 'loóigica', 'lpóigica', 'lqóigica', 'lróigica', 'lsóigica', 'ltóigica', 'luóigica', 'lvóigica', 'lwóigica', 'lxóigica', 'lyóigica', 'lzóigica', 'láóigica', 'lâóigica', 'làóigica', 'lãóigica', 'léóigica', 'lêóigica', 'lèóigica', 'lẽóigica', 'líóigica', 'lîóigica', 'lì

In [32]:
evaluator(test_list)

41.4% of 186 words


There was a clear improvement in the percentage of evaluated words.

### Correcting Typing Errors

In [33]:
def insert_letters(slices):
    new_words = []
    letters = 'abcdefghijklmnopqrstuvwxyzáâàãéêèẽíîìĩóôõòúûùũç'
    for left, right in slices:
        for letter in letters:
            new_words.append(left + letter + right)                             # Inserts the letter between the word
    return new_words

In [34]:
def deleting_characters(slices):
    new_words = []
    for left, right in slices:
        new_words.append(left + right[1:])                                      # On the right side, deletes the second character of the word and returns the second letter onwards
    return new_words

In [35]:
def replace_letter(slices):
    new_words = []
    letters = 'abcdefghijklmnopqrstuvwxyzáâàãéêèẽíîìĩóôõòúûùũç'
    for left, right in slices:
        for letter in letters:
            new_words.append(left + letter + right[1:])                         # On the right side, replaces the second letter of the word and returns the second letter onwards
    return new_words

In [36]:
def invert_letter(slices):
    new_words = []
    for left, right in slices:
        if len(right) > 1:
            new_words.append(left + right[1] + right[0] + right[2:])            # On the right side, inverts the second letter with the first letter and returns the third letter onwards
    return new_words

In [37]:
# Implementing corrections in the word generator
def word_generator(word):
    slices = []
    for i in range(len(word)+1):
        slices.append((word[:i], word[i:]))
    generated_words = insert_letters(slices)
    generated_words += deleting_characters(slices)
    generated_words += replace_letter(slices)
    generated_words += invert_letter(slices)
    return generated_words

In [38]:
evaluator(test_list)
example_word = "lógiac"
generated_words = word_generator(example_word)
print(generated_words)

76.34% of 186 words
['alógiac', 'blógiac', 'clógiac', 'dlógiac', 'elógiac', 'flógiac', 'glógiac', 'hlógiac', 'ilógiac', 'jlógiac', 'klógiac', 'llógiac', 'mlógiac', 'nlógiac', 'ológiac', 'plógiac', 'qlógiac', 'rlógiac', 'slógiac', 'tlógiac', 'ulógiac', 'vlógiac', 'wlógiac', 'xlógiac', 'ylógiac', 'zlógiac', 'álógiac', 'âlógiac', 'àlógiac', 'ãlógiac', 'élógiac', 'êlógiac', 'èlógiac', 'ẽlógiac', 'ílógiac', 'îlógiac', 'ìlógiac', 'ĩlógiac', 'ólógiac', 'ôlógiac', 'õlógiac', 'òlógiac', 'úlógiac', 'ûlógiac', 'ùlógiac', 'ũlógiac', 'çlógiac', 'laógiac', 'lbógiac', 'lcógiac', 'ldógiac', 'leógiac', 'lfógiac', 'lgógiac', 'lhógiac', 'liógiac', 'ljógiac', 'lkógiac', 'llógiac', 'lmógiac', 'lnógiac', 'loógiac', 'lpógiac', 'lqógiac', 'lrógiac', 'lsógiac', 'ltógiac', 'luógiac', 'lvógiac', 'lwógiac', 'lxógiac', 'lyógiac', 'lzógiac', 'láógiac', 'lâógiac', 'làógiac', 'lãógiac', 'léógiac', 'lêógiac', 'lèógiac', 'lẽógiac', 'líógiac', 'lîógiac', 'lìógiac', 'lĩógiac', 'lóógiac', 'lôógiac', 'lõógiac', 'lòógiac', 

In [39]:
evaluator(test_list)

76.34% of 186 words


I obtained better results compared to the previous evaluation which had a slightly higher percentage than 40% of the 186 words.

### Creating the Corrector

In [40]:
# Evaluator with unknown word rate
def evaluator(tests, vocabulary):
    word_count = len(tests)
    correct_count = 0
    unknown = 0
    for correct, wrong in tests:
        corrected_word = corrector(wrong)
        if corrected_word == correct:
            correct_count += 1
        else:
            unknown += (correct not in vocabulary)                              # Counting when the correct word is unknown in my vocabulary
    accuracy_rate = round(correct_count * 100 / word_count, 2)
    unknown_rate = round(unknown * 100 / word_count, 2)                         # Percentage of unknown words

    print(f"{accuracy_rate}% of {word_count} words, unknown is {unknown_rate}%")

vocabulary = set(normalized_list)                                               # Non-repetitive words
evaluator(test_list, vocabulary)

76.34% of 186 words, unknown is 6.99%


So far, the corrector only corrects words that are at a single distance from a correct word. To improve this, I need to enhance my word generator.

In [41]:
word = "lóiigica"

def improved_generator(generated_words):
    new_words = []
    for word in generated_words:
        new_words += word_generator(word)                                       # New words at two distances from corrected words
    return new_words

words_g = improved_generator(word_generator(word))
"lógica" in words_g

True

In [42]:
len(words_g)

787396

In [43]:
def new_corrector(word):
    generated_words = word_generator(word)
    improved_words = improved_generator(generated_words)                        # Includes the generation of improved words
    all_words = set(generated_words + improved_words)                           # Includes the improved words without repetition
    candidates = [word]
    for word in all_words:
        if word in vocabulary:
            candidates.append(word)
    correct_word = max(candidates, key=probability)                             # If the word is contained in the vocabulary, it will return the word with the highest probability of being the correct word
    return correct_word

new_corrector(word)

'lógica'

In [44]:
def evaluator(tests, vocabulary):
    word_count = len(tests)
    correct_count = 0
    unknown = 0

    # Returns the corrected word plus the count of unknown words
    for correct, wrong in tests:
        corrected_word = new_corrector(wrong)
        unknown += (correct not in vocabulary)
        # The condition returns the count of correct words for each correct answer
        if corrected_word == correct:
            correct_count += 1
        else:
            print(wrong + "-" + corrector(wrong) + "-" + corrected_word)

    accuracy_rate = round(correct_count * 100 / word_count, 2)
    unknown_rate = round(unknown * 100 / word_count, 2)

    print(f"{accuracy_rate}% of {word_count} words, unknown is {unknown_rate}%")

vocabulary = set(normalized_list)
evaluator(test_list, vocabulary)

esje-esse-se
sãêo-são-não
dosa-dos-do
eme-em-de
eàssa-essa-esse
daõs-das-da
céda-cada-da
noâ-no-o
enêão-então-não
tĩem-tem-em
nossah-nossa-nosso
teb-tem-de
atĩ-até-a
âem-em-de
foo-foi-o
serr-ser-se
entke-entre-então
van-vai-a
çeus-seus-seu
eû-e-de
temeo-tempo-temos
semre-sempre-ser
elaá-ela-ele
síó-só-se
siàe-site-se
seém-sem-em
peln-pelo-ele
aléra-alura-agora
tdia-dia-da
tuúo-tudo-tipo
jé-é-de
sãô-são-não
odos-dos-do
siua-sua-seu
elpe-ele-esse
teos-temos-os
eũsa-essa-esse
vjmos-vamos-temos
dms-dos-de
cava-java-para
ános-nos-no
èaso-caso-as
túem-tem-em
daáos-dados-dos
nossk-nosso-nosso
tãer-ter-ser
vté-até-é
búm-bem-um
sçerá-será-ser
entró-entre-então
uai-vai-a
sâus-seus-seu
ìeu-seu-de
fual-qual-sua
elal-ela-ele
skó-só-se
secm-sem-em
aluéa-alura-além
dil-dia-de
sód-só-se
eúaa-aeúaa-essa
ró-só-de
dĩaz-adĩaz-da
correptor-corretor-correto
trtica-tática-prática
ewpoderamento-aewpoderamento-ewpoderamento
îgato-gato-fato
cakvalo-acakvalo-carvalho
canelac-acanelac-janela
tênisy-atênisy-tênisy

Considering the line "odos-dos-do" the word "dos" is already correct, however, the code corrects it to "do". Therefore, one of the errors of the improved corrector is associated with the fact that it generates both the possible correct word and an existing variation in the vocabulary, which has a higher frequency. Hence, I will use the consolidated evaluator with the old corrector.

In [45]:
word = "lóiigica"
print(new_corrector(word))
print(corrector(word))

lógica
alóiigica


In [46]:
def evaluator(tests, vocabulary):
    word_count = len(tests)
    correct_count = 0
    unknown = 0
    for correct, wrong in tests:
        corrected_word = corrector(wrong)
        unknown += (correct not in vocabulary)
        if corrected_word == correct:
            correct_count += 1
    accuracy_rate = round(correct_count * 100 / word_count, 2)
    unknown_rate = round(unknown * 100 / word_count, 2)
    print(f"{accuracy_rate}% of {word_count} words, unknown is {unknown_rate}%")

vocabulary = set(normalized_list)
evaluator(test_list, vocabulary)

76.34% of 186 words, unknown is 6.99%
